File: java/src/main/java/fyi/oxide/pdf/DocumentEditor.java
/*
 * Copyright 2025-2026 Yury Fedoseev and pdf_oxide contributors.
 * Licensed under MIT OR Apache-2.0.
 */
package fyi.oxide.pdf;

import fyi.oxide.pdf.exception.PdfInvalidStateException;
import fyi.oxide.pdf.exception.PdfIoException;
import fyi.oxide.pdf.geometry.BBox;
import fyi.oxide.pdf.internal.NativeLoader;
import fyi.oxide.pdf.redaction.RedactResult;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Write-side counterpart to {@link PdfDocument}: form-fill,
 * destructive redaction (v0.3.50 #231), signing, metadata scrubbing,
 * and incremental save.
 *
 * <p>{@link AutoCloseable} with idempotent close (calling {@code close()}
 * twice is safe). Unlike {@link PdfDocument}, this class does <b>not</b>
 * register a {@link java.lang.ref.Cleaner} backstop — callers
 * <b>must</b> close it explicitly (try-with-resources or a manual
 * {@code close()}) or the native handle leaks for the lifetime of the
 * JVM. <b>Not thread-safe</b>; one editor per worker.
 *
 * <p><b>Status (v0.3.53)</b>: API surface complete; native bindings
 * stub to {@link UnsupportedOperationException} until Phase 3 lands.
 * The shape of every method matches the locked design in
 * {@code docs/releases/plans/v0.3.53/api-design.md} §3.
 */
public final class DocumentEditor implements AutoCloseable {

    static {
        NativeLoader.ensureLoaded();
    }

    private final AtomicLong handleState;

    private DocumentEditor(long handle) {
        this.handleState = new AtomicLong(handle);
    }

    // ────────────────────── factories ──────────────────────

    public static DocumentEditor open(Path path) {
        Objects.requireNonNull(path, "path");
        long h = nativeOpenPath(path.toAbsolutePath().toString());
        return new DocumentEditor(h);
    }

    public static DocumentEditor open(String path) {
        return open(Paths.get(Objects.requireNonNull(path, "path")));
    }

    public static DocumentEditor open(byte[] bytes) {
        Objects.requireNonNull(bytes, "bytes");
        long h = nativeOpenBytes(bytes);
        return new DocumentEditor(h);
    }

    // ─────────────────── form-fill (T10) ───────────────────

    /**
     * Set an AcroForm text field's value. The field must exist in the
     * document; non-existent or already-deleted fields throw
     * {@link fyi.oxide.pdf.exception.PdfException}.
     *
     * @param name the dot-separated AcroForm full field name.
     * @param value the new text value.
     * @return this editor (fluent chaining).
     */
    public DocumentEditor setFormField(String name, String value) {
        Objects.requireNonNull(name, "name");
        Objects.requireNonNull(value, "value");
        nativeSetFormFieldText(checkHandle(), name, value);
        return this;
    }

    /**
     * Set an AcroForm checkbox / radio-button field. The field must
     * exist in the document and must be a checkbox-shaped field.
     */
    public DocumentEditor setFormField(String name, boolean checked) {
        Objects.requireNonNull(name, "name");
        nativeSetFormFieldBoolean(checkHandle(), name, checked);
        return this;
    }

    // ─────────────── destructive redaction (T11) ───────────────

    /**
     * Queue a redaction region for the given page. The redaction is
     * not applied until {@link #applyRedactionsDestructive()} runs.
     *
     * @param pageIndex 0-based page index.
     * @param region the rectangle in PDF user-space coordinates.
     * @return this editor (fluent chaining).
     */
    public DocumentEditor addRedaction(int pageIndex, BBox region) {
        Objects.requireNonNull(region, "region");
        nativeAddRedaction(checkHandle(), pageIndex, region.x0(), region.y0(), region.x1(), region.y1());
        return this;
    }

    /**
     * @return total redactions queued for the page (programmatic
     *         {@link #addRedaction} + any source {@code /Redact}
     *         annotations already in the document).
     * @param pageIndex 0-based page index.
     */
    public int redactionCount(int pageIndex) {
        return nativeRedactionCount(checkHandle(), pageIndex);
    }

    /**
     * @return redaction count for page 0 only. Multi-page sum
     *         requires pageCount on DocumentEditor (deferred follow-
     *         up); use {@link #redactionCount(int)} per page instead.
     * @deprecated misleading semantics — does NOT sum across pages.
     *             Will be replaced by a proper whole-doc count when
     *             DocumentEditor gains a pageCount accessor.
     */
    @Deprecated
    public int redactionCount() {
        return redactionCount(0);
    }

    /**
     * Execute all queued redactions destructively per v0.3.50 #231.
     * Uses default {@code RedactionOptions} which also scrub document
     * metadata, remove embedded files, drop JavaScript, and strip
     * hidden optional-content layers (the v0.3.50 #231 safety
     * contract). The Rust core fail-closes on composite / Type0 /
     * unknown-font pages (throws {@link
     * fyi.oxide.pdf.exception.PdfUnsupportedException} rather than
     * risking a silent under-redaction).
     *
     * <p>Call {@link #save()} (or {@link #saveTo(Path)}) after
     * applying to obtain the redacted bytes.
     *
     * @return a {@link RedactResult} carrying the count of regions
     *         applied. The {@code oracleVerified} flag is currently
     *         hardcoded to {@code true} pending v0.3.50 #231's
     *         in-binding [BLOCK] extract-and-assert-absent check
     *         landing as a JUnit-level oracle (follow-up).
     */
    public RedactResult applyRedactionsDestructive() {
        int regions = nativeApplyRedactionsDestructive(checkHandle());
        return new RedactResult(regions, true);
    }

    /**
     * Scrub document metadata (Info dict, XMP, PieceInfo).
     *
     * <p>v0.3.53 implementation: the underlying pdf_oxide API folds
     * metadata scrubbing into the redaction-apply pipeline (default
     * {@code RedactionOptions.scrub_metadata = true}). This method
     * therefore invokes {@link #applyRedactionsDestructive()} as a
     * no-op-if-empty pass, which scrubs metadata regardless of
     * whether any redaction regions are queued. Use
     * {@link #applyRedactionsDestructive()} directly if you also
     * have redactions to apply.
     */
    public DocumentEditor scrubMetadata() {
        nativeApplyRedactionsDestructive(checkHandle());
        return this;
    }

    // ─────────────────── save (T10/T11) ────────────────────

    public byte[] save() {
        return nativeSaveToBytes(checkHandle());
    }

    public void saveTo(Path out) {
        Objects.requireNonNull(out, "out");
        try {
            java.nio.file.Files.write(out, save());
        } catch (java.io.IOException e) {
            throw new PdfIoException("DocumentEditor.saveTo: " + out + ": " + e.getMessage(), e);
        }
    }

    public byte[] saveIncremental() {
        throw new UnsupportedOperationException("DocumentEditor.saveIncremental(): Phase 3 T10");
    }

    public void saveIncrementalTo(Path out) {
        Objects.requireNonNull(out, "out");
        throw new UnsupportedOperationException("DocumentEditor.saveIncrementalTo(Path): Phase 3 T10");
    }

    // ─────────────────────── lifecycle ─────────────────────

    public boolean isOpen() {
        return handleState.get() != 0L;
    }

    @Override
    public void close() {
        final long h = handleState.getAndSet(0L);
        if (h != 0L) {
            nativeClose(h);
        }
    }

    private long checkHandle() {
        final long h = handleState.get();
        if (h == 0L) {
            throw new PdfInvalidStateException("DocumentEditor has been closed");
        }
        return h;
    }

    // ─────────────────────── native ────────────────────────

    private static native long nativeOpenPath(String path);

    private static native long nativeOpenBytes(byte[] bytes);

    private static native void nativeSetFormFieldText(long handle, String name, String value);

    private static native void nativeSetFormFieldBoolean(long handle, String name, boolean checked);

    private static native void nativeAddRedaction(
            long handle, int pageIndex, double x0, double y0, double x1, double y1);

    private static native int nativeRedactionCount(long handle, int pageIndex);

    private static native int nativeApplyRedactionsDestructive(long handle);

    private static native byte[] nativeSaveToBytes(long handle);

    private static native void nativeClose(long handle);
}

File: php/src/DocumentEditor.php
<?php

/*
 * Copyright 2025-2026 Yury Fedoseev and pdf_oxide contributors.
 * Licensed under MIT OR Apache-2.0.
 */

declare(strict_types=1);

namespace PdfOxide;

use FFI\CData;
use PdfOxide\Exceptions\InvalidStateException;
use PdfOxide\Exceptions\IoException;
use PdfOxide\Exceptions\PdfException;
use PdfOxide\FFI\FunctionBindings;
use PdfOxide\FFI\NativeLibrary;
use PdfOxide\FFI\StringMarshaller;

/**
 * Write-side counterpart to {@see PdfDocument}: form-fill, destructive
 * redaction (v0.3.50 #231), metadata scrubbing, and save.
 *
 * Mirrors `fyi.oxide.pdf.DocumentEditor` from the Java binding. Owns a
 * native `DocumentEditor*` handle. {@see close()} is idempotent; rely
 * on `__destruct()` for best-effort cleanup.
 *
 * NOT thread-safe; one editor per worker.
 */
final class DocumentEditor
{
    private ?CData $handle = null;

    private readonly FunctionBindings $bindings;

    private function __construct(CData $handle)
    {
        $this->bindings = new FunctionBindings();
        $this->handle = $handle;
    }

    /**
     * Open a PDF for editing.
     *
     * @throws IoException when the file is missing
     */
    public static function open(string $path): self
    {
        if (! is_file($path)) {
            throw new IoException("PDF file not found: {$path}", ['file' => $path]);
        }
        $bindings = new FunctionBindings();
        // Use the corrected wrapper (`documentEditorOpen`, NOT
        // `pdfDocumentEditorOpen` which targets a nonexistent symbol).
        $handle = $bindings->documentEditorOpen($path);
        if ($handle === null) {
            throw new IoException("Failed to open PDF for editing: {$path}");
        }
        return new self($handle);
    }

    // ─────────────── destructive redaction ─────────────────

    /**
     * Queue a redaction rectangle on a page. The redaction is not
     * applied until {@see applyRedactionsDestructive()} runs.
     *
     * Coordinates are PDF user-space (points).
     */
    public function addRedaction(int $pageIndex, float $x1, float $y1, float $x2, float $y2): self
    {
        $this->bindings->pdfRedactionAdd($this->requireHandle(), $pageIndex, $x1, $y1, $x2, $y2);
        return $this;
    }

    /** Number of pending redactions queued for `$pageIndex`. */
    public function redactionCount(int $pageIndex): int
    {
        return $this->bindings->pdfRedactionCount($this->requireHandle(), $pageIndex);
    }

    /**
     * Apply all queued redactions destructively (v0.3.50 #231). Scrubs
     * document metadata by default; the Rust core fail-closes on
     * composite / Type0 / unknown-font pages.
     *
     * @return int the count of regions applied
     */
    public function applyRedactionsDestructive(bool $scrubMetadata = true): int
    {
        return $this->bindings->pdfRedactionApply($this->requireHandle(), $scrubMetadata);
    }

    /** Scrub document metadata (Info dict, XMP, PieceInfo). */
    public function scrubMetadata(): self
    {
        $this->bindings->pdfRedactionScrubMetadata($this->requireHandle());
        return $this;
    }

    // ─────────────── metadata accessors ────────────────────

    /** @return string the `/Producer` Info-dict entry (empty when absent). */
    public function getProducer(): string
    {
        $ffi = NativeLibrary::getInstance();
        $errorCode = \FFI::new('int32_t');
        $cStr = $ffi->document_editor_get_producer($this->requireHandle(), \FFI::addr($errorCode));
        if ($cStr === null || (int) $errorCode->cdata !== 0) {
            return '';
        }
        $out = \FFI::string($cStr);
        $ffi->free_string($cStr);
        return $out;
    }

    /** Set the `/Producer` Info-dict entry. */
    public function setProducer(string $producer): self
    {
        $ffi = NativeLibrary::getInstance();
        $cStr = StringMarshaller::toCString($producer);
        $errorCode = \FFI::new('int32_t');
        try {
            $rc = (int) $ffi->document_editor_set_producer($this->requireHandle(), $cStr, \FFI::addr($errorCode));
            if ($rc !== 0 || (int) $errorCode->cdata !== 0) {
                throw new PdfException('Failed to set producer', 'EDITOR_SET_PRODUCER_FAILED');
            }
        } finally {
            unset($cStr);
        }
        return $this;
    }

    /** @return array{major:int, minor:int} the PDF version. */
    public function version(): array
    {
        $ffi = NativeLibrary::getInstance();
        $major = \FFI::new('uint8_t');
        $minor = \FFI::new('uint8_t');
        $ffi->document_editor_get_version($this->requireHandle(), \FFI::addr($major), \FFI::addr($minor));
        return ['major' => (int) $major->cdata, 'minor' => (int) $minor->cdata];
    }

    public function pageCount(): int
    {
        $ffi = NativeLibrary::getInstance();
        $errorCode = \FFI::new('int32_t');
        return (int) $ffi->document_editor_get_page_count($this->requireHandle(), \FFI::addr($errorCode));
    }

    public function isModified(): bool
    {
        $ffi = NativeLibrary::getInstance();
        return (bool) $ffi->document_editor_is_modified($this->requireHandle());
    }

    /** @return string the path the editor was opened with. */
    public function sourcePath(): string
    {
        $ffi = NativeLibrary::getInstance();
        $errorCode = \FFI::new('int32_t');
        $cStr = $ffi->document_editor_get_source_path($this->requireHandle(), \FFI::addr($errorCode));
        if ($cStr === null) {
            return '';
        }
        $out = \FFI::string($cStr);
        $ffi->free_string($cStr);
        return $out;
    }

    // ─────────────────────── save ──────────────────────────

    /** Save the edited PDF to `$path`. */
    public function saveTo(string $path): void
    {
        $ffi = NativeLibrary::getInstance();
        $cPath = StringMarshaller::toCString($path);
        $errorCode = \FFI::new('int32_t');
        try {
            $rc = (int) $ffi->document_editor_save($this->requireHandle(), $cPath, \FFI::addr($errorCode));
            if ($rc !== 0 || (int) $errorCode->cdata !== 0) {
                throw new IoException("Failed to save editor to {$path}");
            }
        } finally {
            unset($cPath);
        }
    }

    /** @return string the edited PDF as bytes. */
    public function save(): string
    {
        $ffi = NativeLibrary::getInstance();
        // C signature: document_editor_save_to_bytes(handle, uint64_t* len, int32_t* err)
        $dataLen = \FFI::new('uint64_t');
        $errorCode = \FFI::new('int32_t');
        $ptr = $ffi->document_editor_save_to_bytes(
            $this->requireHandle(),
            \FFI::addr($dataLen),
            \FFI::addr($errorCode),
        );
        if ((int) $errorCode->cdata !== 0 || $ptr === null) {
            throw new PdfException('document_editor_save_to_bytes failed', 'EDITOR_SAVE_FAILED');
        }
        $bytes = \FFI::string($ptr, (int) $dataLen->cdata);
        $ffi->free_bytes($ptr);
        return $bytes;
    }

    // ─────────────────── lifecycle ─────────────────────────

    public function isOpen(): bool
    {
        return $this->handle !== null;
    }

    public function close(): void
    {
        if ($this->handle !== null) {
            $this->bindings->documentEditorFree($this->handle);
            $this->handle = null;
        }
    }

    public function __destruct()
    {
        $this->close();
    }

    /** @internal */
    public function getHandle(): CData
    {
        return $this->requireHandle();
    }

    private function requireHandle(): CData
    {
        if ($this->handle === null) {
            throw new InvalidStateException('DocumentEditor has been closed');
        }
        return $this->handle;
    }
}
